File: TestLeetCode/SearchInRotatedArray.cs
using System;

public class SearchRotatedArraySolution {
    public int Search(int[] nums, int target) {
        int l = 0, r = nums.Length - 1;

        while (l <= r) {
            int mid = (l + r) / 2;
            if (target == nums[mid]) {
                return mid;
            }

            // Left sorted portion
            if (nums[l] <= nums[mid]) {
                if (target > nums[mid] || target < nums[l]) {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
            // Right sorted portion
            else {
                if (target < nums[mid] || target > nums[r]) {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }
        }
        return -1;
    }
}

File: python/search/search_in_rotated_array.py
class RotatedArraySearch:
    def search(self, numbers: list[int], target: int) -> int:
        left, right = 0, len(numbers) - 1
        while left <= right:
            middle = (left + right) // 2
            if numbers[middle] == target: return middle
            if numbers[left] <= numbers[middle]:
                if numbers[left] <= target < numbers[middle]: right = middle - 1
                else: left = middle + 1
            elif numbers[middle] < target <= numbers[right]: left = middle + 1
            else: right = middle - 1
        return -1
